builtin/logical/pki/crl_util.go

```diff
@@ -422,6 +447,187 @@ func (cb *crlBuilder) rebuildDeltaCRLsHoldingLock(sc *storageContext, forceNew b
+		return nil
+	}
+
+	sc.Backend.Logger().Debug(fmt.Sprintf("gathering first time existing revocations"))
+
+	clusters, err := sc.Storage.List(sc.Context, crossRevocationPrefix)
+	if err != nil {
+		return fmt.Errorf("failed to list cross-cluster revocation queue participating clusters: %w", err)
+	}
+
+	sc.Backend.Logger().Debug(fmt.Sprintf("found %v clusters: %v", len(clusters), clusters))
+
+	for cIndex, cluster := range clusters {
+		cluster = cluster[0 : len(cluster)-1]
+		cPath := crossRevocationPrefix + cluster + "/"
+		serials, err := sc.Storage.List(sc.Context, cPath)
+		if err != nil {
+			return fmt.Errorf("failed to list cross-cluster revocation queue entries for cluster %v (%v): %w", cluster, cIndex, err)
+		}
+
+		sc.Backend.Logger().Debug(fmt.Sprintf("found %v serials for cluster %v: %v", len(serials), cluster, serials))
+
+		for _, serial := range serials {
+			if serial[len(serial)-1] == '/' {
+				serial = serial[0 : len(serial)-1]
+			}
+
+			ePath := cPath + serial
+			eConfirmPath := ePath + "/confirmed"
+			removalEntry, err := sc.Storage.Get(sc.Context, eConfirmPath)
+
+			entry := &revocationQueueEntry{
+				Cluster: cluster,
+				Serial:  serial,
+			}
+
+			// No removal entry yet; add to regular queue. Otherwise, slate it
+			// for removal if we're a perfPrimary.
+			if err != nil || removalEntry == nil {
+				cb.revQueue.Add(entry)
+			} else if !isNotPerfPrimary {
+				cb.removalQueue.Add(entry)
+			} else {
+				sc.Backend.Logger().Debug(fmt.Sprintf("ignoring confirmed revoked serial %v: %v vs %v ", serial, err, removalEntry))
+			}
+
+			// Overwrite the error; we don't really care about its contents
+			// at this step.
+			err = nil
+		}
+	}
+
+	return nil
+}
+
+func (cb *crlBuilder) processRevocationQueue(sc *storageContext) error {
+	sc.Backend.Logger().Debug(fmt.Sprintf("starting to process revocation requests"))
+
+	isNotPerfPrimary := sc.Backend.System().ReplicationState().HasState(consts.ReplicationDRSecondary|consts.ReplicationPerformanceStandby) ||
+		(!sc.Backend.System().LocalMount() && sc.Backend.System().ReplicationState().HasState(consts.ReplicationPerformanceSecondary))
+
+	// Before revoking certificates, we need to hold the lock for certificate
+	// storage. This prevents any parallel revocations and prevents us from
+	// multiple places. We do this before grabbing the contents of the
+	// revocation queues themselves, to ensure we interleave well with other
+	// invocations of this function and avoid duplicate work.
+	sc.Backend.revokeStorageLock.Lock()
+	defer sc.Backend.revokeStorageLock.Unlock()
+
+	if err := cb.maybeGatherQueueForFirstProcess(sc, isNotPerfPrimary); err != nil {
+		return fmt.Errorf("failed to gather first queue: %v", err)
+	}
+
+	revQueue := cb.revQueue.Iterate()
+	removalQueue := cb.removalQueue.Iterate()
+
+	sc.Backend.Logger().Debug(fmt.Sprintf("gathered %v revocations and %v confirmation entries", len(revQueue), len(removalQueue)))
+
+	for _, req := range revQueue {
```

Comment:
Follow-up PR can ignore any revocation requests that occur on our cluster, due to now knowing `clusterId`.